### PR TITLE
Update to Berkeley DB 5.3

### PR DESCRIPTION
**Pull request by rnicoll**
Alternative to #2221

### CODE REVIEW

**Comment by rnicoll** [comment]
Failure does not appear on my machine, retrying in the hope it was a random infrastructure hiccup.

**Comment by rnicoll** [comment]
Okay, next step is to recreate on a CentOS 7 VM

**Comment by rnicoll** [comment]
Can't even get the code to compile on CentOS (looks like getting the right boost packages is tricky). Trying again to reproduce the issue locally.

**Comment by patricklodder** [comment]
I'm scheduling this for cross-compatibility testing AFTER 1.21-dev is capable of fully synching the (testnet3) chain (i.e. AuxPoW works) so I am marking this with "do not merge yet" label until then. Please, do not merge yet ❤️

**Comment by bjacquin** [comment]
See https://github.com/dogecoin/dogecoin/pull/2483 for another version focusing solely on DB 5.3

**Comment by patricklodder** [comment]
@rnicoll currently I'm having issues with 1.21 expecting bdb4.8 whereas 1.14.4+ already has 5.3. Would it be possible to move this one forward first, or do we still have other work that you'd want in before this?

**Comment by rnicoll** [comment]
@patricklodder Yeah, can do. Rebasing it without the interim pieces, now, and will then see how it looks.

**Comment by rnicoll** [comment]
@bjacquin Yup, using your patch on top of mine. Looks like we've got a few CI cases which are harder to adapt (i.e. Homebrew ships 4.8 and 18.x, but nothing in between) but most of the way there.

**Comment by patricklodder** [comment]
@rnicoll Re: 4.8 for OSX-native. That's a bit of a portability issue for the wallet.dat. Don't want to make it blocking, but should be on a TODO for 1.21, imho.

Potential solution: @michilumin made a guide for 1.14.2 that I recently proposed to make standard for 1.14, at #2686. After we merge this, it may be an idea to just use the methodology from the guide (download bdb5.3 from source and compile it) for the system-lib-built OSX test. It also makes it easier to have that guide, because the CI will say if it's still good 😊

**Comment by rnicoll** [comment]
@patricklodder Yeah... trying to get the clang patch to reliably apply (because it seemed to skip on exactly one CI target), then will take another look at OS X with a view of it downloading just that dependency.

**Comment by rnicoll** [comment]
@patricklodder Looks like there's no ideal solution, as BDB 5.3 not working on arm64 is an ongoing issue: https://trac.macports.org/ticket/62216

Going to try patching BDB to compile on x86 for now and we can think about arm64 later.

**Comment by patricklodder** [comment]
> BDB 5.3 not working on arm64

Hmm it works for other arm64...

I'll try to do it on my M1 and let you know what I end up with. Can fix that in a subsequent PR.

**Comment by rnicoll** [comment]
@patricklodder Going to throw this through Gitian now just to be sure, and do a few upgrade checks. Then do we want to just disable 10.14 native and patch in a follow-on, or leave until we figure out 10.14?

**Comment by rnicoll** [comment]
Turns out it doesn't build on Gitian, but it doesn't build on Gitian before this. Both of which are strange because I was damn sure I'd tried 1.21 on Gitian :-/

**Comment by patricklodder** [comment]
> Then do we want to just disable 10.14 native and patch in a follow-on, or leave until we figure out 10.14?

I'd even be fine with keeping the 10.14 build with 4.8 and then we fix it in a separate pr. I think most testing right now happens on linux?

> Turns out it doesn't build on Gitian, but it doesn't build on Gitian before this.

We can fix Gitian later, maybe file an issue with the error?

**Comment by patricklodder** [comment]
### Interim report from native M1 bdb53 building

> Looks like there's no ideal solution, as BDB 5.3 not working on arm64 is an ongoing issue: https://trac.macports.org/ticket/62216

I get around this issue with:

```bash
cd build_unix
../dist/configure CXX=clang++ --enable-cxx --enable-posixmutexes
```

I get some compiler errors after that, which I am investigating later today, but this specific issue seems solvable.

File: ci/test/05_before_script.sh [hunk 1]
```diff
@@ -25,10 +25,15 @@ if [ -n "$XCODE_VERSION" ] && [ ! -f "$OSX_SDK_PATH" ]; then
...
+  # Use BDB compiled using install_db5.sh script to work around linking issue when using BDB
   # from depends. See https://github.com/bitcoin/bitcoin/pull/18288#discussion_r433189350 for
   # details.
-  DOCKER_EXEC "contrib/install_db4.sh \$(pwd) --enable-umrw CC=clang CXX=clang++ CFLAGS='${MSAN_FLAGS}' CXXFLAGS='${MSAN_AND_LIBCXX_FLAGS}'"
+  DOCKER_EXEC "contrib/install_db5.sh \$(pwd) --enable-umrw CC=clang CXX=clang++ CFLAGS='${MSAN_FLAGS}' CXXFLAGS='${MSAN_AND_LIBCXX_FLAGS}'"
+fi
+
+if [[ $HOST = *-apple-darwin16 ]]; then
+  # Use BDB compiled using install_db5.sh script as Brew does not supply BDB 5.3
+  DOCKER_EXEC "contrib/install_db5.sh \$(pwd)"
```
Comment (patricklodder) [inline ci/test/05_before_script.sh:36]: locally, adding `--enable-posixmutexes` was all I needed to make this work on my M1:

```console
% clang --version
Apple clang version 12.0.5 (clang-1205.0.22.11)
Target: arm64-apple-darwin20.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

```suggestion
  DOCKER_EXEC "contrib/install_db5.sh \$(pwd) --enable-posixmutexes"
```

Comment (rnicoll) [reply ci/test/05_before_script.sh]: That did seem a lot better, although now it's not finding the library, for reasons I can't readily see :-/

Comment (patricklodder) [reply ci/test/05_before_script.sh]: I think that this is similar to an m4 thing I PR'd against #2745 - but in a different spot. I still have PR-ing you a fix on my todo. I started but had some issues with cirrus. Will come back to that.

**Comment by patricklodder** [comment]
Merged through #2885

**Pull request closed without merging**